caffe2/core/nomnigraph/Converters/Caffe2.cc

```diff
@@ -288,6 +316,62 @@ repr::NNModule convertFromCaffe2Proto(const caffe2::NetDef &net, std::unordered_
   return module;
 }
 
+caffe2::OperatorDef convertOperator(repr::NNGraph::NodeRef instrNode) {
+  auto *nnOp = repr::nn::get<repr::NeuralNetOperator>(instrNode);
+  auto *annotation = nnOp->getAnnotation();
+  caffe2::OperatorDef op;
+
+  if (!annotation || !annotation->getSaved()) {
+    using NNKind = repr::NeuralNetOperator::NNKind;
+    switch (nnOp->getKind()) {
+      case NNKind::Send:
+      {
+        auto sendOp = dyn_cast<repr::Send>(nnOp);
+        op.set_type("Placeholder:Send");
+        {
+          auto arg = op.add_arg();
+          arg->set_name("dst_node");
+          arg->set_s(sendOp->getDestination());
+        }
+        {
+          auto out = repr::nn::getOutputs(instrNode).front();
+          auto t = repr::nn::get<repr::Data>(out);
+          auto arg = op.add_arg();
+          arg->set_name("callsite_id");
+          arg->set_i(t->getVersion());
+        }
+        break;
+      }
+      case NNKind::Receive:
+      {
+        auto recvOp = dyn_cast<repr::Receive>(nnOp);
+        op.set_type("Placeholder:Receive");
+        {
+          auto arg = op.add_arg();
+          arg->set_name("src_node");
+          arg->set_s(recvOp->getSource());
+        }
+        {
+          auto out = repr::nn::getOutputs(instrNode).front();
+          auto t = repr::nn::get<repr::Data>(out);
+          auto arg = op.add_arg();
+          arg->set_name("callsite_id");
+          arg->set_i(t->getVersion());
+        }
+        break;
+      }
+      default:
+        op.set_type(nnOp->getName());
+    }
+  } else {
+    op = *reinterpret_cast<caffe2::OperatorDef *>(annotation->getSaved());
```

Comment:
What is going on here?

Reply:
annotation->getSaved() gives a generic void* that needs to be interpretted as an operatordef pointer.  I'll migrate this entire file into caffe2/caffe2/opt/ (which will enable CAFFE_ENFORCE etc) and create a Caffe2Annotation : Annotation class that exposes this without the need for a generic void* reinterpret_cast